src/MultiFormats/Cryptography/KeccakManaged.cs

```diff
@@ -0,0 +1,332 @@
+
+using System;
+
+namespace MultiFormats.Cryptography
+{
+    internal partial class KeccakManaged : Keccak
+    {
+        public KeccakManaged(int hashBitLength)
+            : base(hashBitLength) { }
+
+        protected override void HashCore(byte[] array, int ibStart, int cbSize)
+        {
+            base.HashCore(array, ibStart, cbSize);
+            if (cbSize == 0)
+                return;
+            var sizeInBytes = SizeInBytes;
+            if (buffer == null)
+                buffer = new byte[sizeInBytes];
+            var stride = sizeInBytes >> 3;
+            var utemps = new ulong[stride];
+            if (buffLength == sizeInBytes)
+                throw new Exception("Unexpected error, the internal buffer is full");
+            AddToBuffer(array, ref ibStart, ref cbSize);
+            if (buffLength == sizeInBytes) //buffer full
+            {
+                Buffer.BlockCopy(buffer, 0, utemps, 0, sizeInBytes);
+                KeccakF(utemps, stride);
+                buffLength = 0;
+            }
+
+            for (; cbSize >= sizeInBytes; cbSize -= sizeInBytes, ibStart += sizeInBytes)
+            {
+                Buffer.BlockCopy(array, ibStart, utemps, 0, sizeInBytes);
+                KeccakF(utemps, stride);
+            }
+
+            if (cbSize > 0) //some left over
+            {
+                Buffer.BlockCopy(array, ibStart, buffer, buffLength, cbSize);
+                buffLength += cbSize;
+            }
+        }
+
+        protected override byte[] HashFinal()
+        {
+            var sizeInBytes = SizeInBytes;
+            var outb = new byte[HashByteLength];
+
+            //    padding
+            if (buffer == null)
+                buffer = new byte[sizeInBytes];
+            else
+                Array.Clear(buffer, buffLength, sizeInBytes - buffLength);
+            buffer[buffLength++] = 1;
+            buffer[sizeInBytes - 1] |= 0x80;
+            var stride = sizeInBytes >> 3;
+            var utemps = new ulong[stride];
+            Buffer.BlockCopy(buffer, 0, utemps, 0, sizeInBytes);
+            KeccakF(utemps, stride);
+            Buffer.BlockCopy(state, 0, outb, 0, HashByteLength);
+            return outb;
+        }
+
+        private void KeccakF(ulong[] inb, int laneCount)
+        {
+            while (--laneCount >= 0)
+                state[laneCount] ^= inb[laneCount];
+            ulong Aba, Abe, Abi, Abo, Abu;
+            ulong Aga, Age, Agi, Ago, Agu;
+            ulong Aka, Ake, Aki, Ako, Aku;
+            ulong Ama, Ame, Ami, Amo, Amu;
+            ulong Asa, Ase, Asi, Aso, Asu;
+            ulong BCa, BCe, BCi, BCo, BCu;
+            ulong Da, De, Di, Do, Du;
+            ulong Eba, Ebe, Ebi, Ebo, Ebu;
+            ulong Ega, Ege, Egi, Ego, Egu;
+            ulong Eka, Eke, Eki, Eko, Eku;
+            ulong Ema, Eme, Emi, Emo, Emu;
+            ulong Esa, Ese, Esi, Eso, Esu;
+            var round = laneCount;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this useless assignment to local variable 'round'.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)